src/js/models/MenuObject.js

```diff
@@ -0,0 +1,251 @@
+import axios from 'axios'
+
+/**
+ * A base data model for menu items,
+ * including relevant business logic.
+ * 
+ * It stores keys-value pairs for each subcomponent,
+ * and an array of those keys for sorting.
+ * 
+ * It contains several methods that can be overridden
+ * to ensure reactivity in whatever framework might
+ * be in use.
+ * 
+ * The basic structure is:
+ * {
+ *    id: int,
+ *    hash: String, // For sync lookups later
+ *    items: {
+ *              1: {attrs...},
+ *              2: {attrs...},
+ *              3: {attrs...},
+ *              ...,
+ *           },
+ *    itemList: [1,3,2, ...],
+ * }
+ */
+class MenuObject {
+  constructor(sessionID, user, set, itemIDKey, ajaxPayload) {
+    this.sessionID = sessionID // axios post functionality
+    this.user = user
+    this.set = set || ((object, key, value) => { object[key] = value }) // This allows you to pass a custom data setter [for proper reactivity]
+    this.itemIDKey = itemIDKey // This is must match the key of the UID returned from the database for this data item (version_id for combinations, image_catalog_id for images, etc.)
+    this._ajaxPayload = Object.assign(
+      {}, 
+      ajaxPayload, 
+      {SESSION_ID: this.sessionID}
+    )
+
+    this._hash = ''
+    this._items = {}
+    this._itemList = []
+  }
+
+  /**
+   * Destroy and clean up memory
+   * 
+   * @public
+   * @instance
+   */
+  destroy() {
+    for (var i = 0, n = this._itemList.length; i < n; i++) {
+      this._itemList[i].destroy()
+      delete this._itemList[i]
+    }
+    delete this._itemList
+
+    for (var key in this._items) {
+        if (this._items.hasOwnProperty(key)) {
+            this._items[key].destroy()
+            delete this._items[key]
+        }
+    }
+    delete this._items
+
+    delete this._hash
+    delete this._id
+  }
+
+  /**
+   * @public 
+   * @instance
+   * 
+   * @return {Array<Number>}       list of all items
+   */
+  items() {
+    return (this._itemList)
+  }
+
+  /**
+   * @public 
+   * @instance
+   * 
+   * @param {number} index the item index to retrieve 
+   * 
+   * @return {Model}       the Model object
+   */
+  itemAtIndex(index) {
+    return (this._items[this._itemList[index]] || null)
+  }
+
+  /**
+   * @public 
+   * @instance
+   * 
+   * @param {number} index the item index to retrieve 
+   * 
+   * @return {Model}       the Model object
+   */
+  itemWithID(id) {
```

Comment:
I have two things going on here.  There is a javascript object _items containing all the items as key-value pairs, then there is an array _itemsList that contains all those keys in a particular order.  So if I want to interact with the ordering of my key-value pairs, I just change the order in the array _itemList (I can also add/remove items that way), all of the actual data is stored in the _items javascript object, and because it is an object, not an array, the ordering there is totally irrelevant.  I picked this model up when reading some articles on Vuex and using these sorts of data models in reactive templates.  I would love to hear your experience and thoughts.  We should discuss data models in any event and perhaps come to a consistent approach.

Reply:
Interesting, yeah, worthy of discussion. There's certainly merit having an object that you can reference by key--it's faster to perform lookups that way. The problem then comes if ordering is also needful ... you have to apply an order from the outside, which has it's own performance pains. Btw, checkout ordered sets in immutable: https://facebook.github.io/immutable-js/docs/#/OrderedSet

Reply:
The ordered set looks really nice.  I think I will give Immutable a shot, it shouldn't be too hard to implement.